### PR TITLE
Active active per replica journal selection

### PR DESCRIPTION
**Pull request by johanandren**
References #29257

### CODE REVIEW

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ActiveActiveShardingDirectReplication.scala [hunk 1]
```diff
@@ -53,15 +54,17 @@ object ActiveActiveShardingDirectReplication {
...
+      replicaShardingProxies: java.util.Map[ReplicaId, ActorRef[T]]): Behavior[Command] =
     apply(selfReplica, replicaShardingProxies.asScala.toMap)
 
   /**
    * Scala API:
    * @param selfReplica The replica id of the replica that runs on this node
    * @param replicaShardingProxies A replica id to sharding proxy mapping for each replica in the system
    */
-  def apply[T](selfReplica: String, replicaShardingProxies: Map[String, ActorRef[T]]): Behavior[Command] =
+  def apply[T](selfReplica: ReplicaId, replicaShardingProxies: Map[ReplicaId, ActorRef[T]]): Behavior[Command] =
```
Comment (chbatey) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ActiveActiveShardingDirectReplication.scala:67]: 👍

File: akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/InMemStorageExtension.scala [hunk 2]
```diff
@@ -4,26 +4,55 @@
...
+  def defaultStorage(): EventStorage = storageFor(PersistenceTestKitPlugin.PluginId)
+
+  // shortcuts for default policy
+  def currentPolicy: ProcessingPolicy[JournalOperation] = defaultStorage().currentPolicy
+  def setPolicy(policy: ProcessingPolicy[JournalOperation]): Unit = defaultStorage().setPolicy(policy)
+  def resetPolicy(): Unit = defaultStorage().resetPolicy()
+
+  def storageFor(key: String): EventStorage =
+    stores.computeIfAbsent(key, _ => {
+      // we don't really care about the key here, we just want separate instances
```
Comment (chbatey) [inline akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/InMemStorageExtension.scala:50]: Gotcha

File: akka-persistence-testkit/src/main/scala/akka/persistence/testkit/query/scaladsl/PersistenceTestKitReadJournal.scala [hunk 3]
```diff
@@ -11,17 +11,28 @@ import akka.persistence.testkit.EventStorage
...
+import com.github.ghik.silencer.silent
+import com.typesafe.config.Config
+import org.slf4j.LoggerFactory
 
 object PersistenceTestKitReadJournal {
   val Identifier = "akka.persistence.testkit.query"
 }
 
-final class PersistenceTestKitReadJournal(system: ExtendedActorSystem)
+@silent("value config") // is never used
```
Comment (chbatey) [inline akka-persistence-testkit/src/main/scala/akka/persistence/testkit/query/scaladsl/PersistenceTestKitReadJournal.scala:22]: notused annotation instead

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ActiveActiveEventSourcing.scala [hunk 4]
```diff
@@ -92,21 +97,43 @@ object ActiveActiveEventSourcing {
...
-   * @return
+   * @param queryPluginId A single query plugin used to read the events from other replicas. Must be the query side of your configured journal plugin.
    */
-  def apply[Command, Event, State](
+  def withSharedJournal[Command, Event, State](
       entityId: String,
-      replicaId: String,
-      allReplicaIds: Set[String],
+      replicaId: ReplicaId,
+      allReplicaIds: Set[ReplicaId],
```
Comment (chbatey) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ActiveActiveEventSourcing.scala:108]: so the map vs set overloading didn't work here?

**Review by chbatey — APPROVED** [review]
LGTM, +1 to just programatic for now

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ActiveActiveEventSourcing.scala (hunk 4 above)
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ActiveActiveEventSourcing.scala:108]: Not sure what that refers to

Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/ActiveActiveEventSourcing.scala:108]: Ah, yes, got a name clash, I think it is related to having two parameter lists but not sure.

**Pull request merged**